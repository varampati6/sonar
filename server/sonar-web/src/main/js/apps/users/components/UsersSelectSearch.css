.Select-big .Select-control {
  padding-top: 4px;
  padding-bottom: 4px;
}

.Select-big .Select-placeholder {
  margin-top: 4px;
  margin-bottom: 4px;
}

.Select-big .Select-value-label {
  margin-top: 5px;
}
