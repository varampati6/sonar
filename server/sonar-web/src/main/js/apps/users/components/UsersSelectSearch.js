/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
//@flow
import React from 'react';
import Select from 'react-select';
import { debounce } from 'lodash';
import { translate, translateWithParameters } from '../../../helpers/l10n';
import UsersSelectSearchOption from './UsersSelectSearchOption';
import UsersSelectSearchValue from './UsersSelectSearchValue';
import './UsersSelectSearch.css';

export type Option = {
  login: string,
  name: string,
  email?: string,
  avatar?: string,
  groupCount?: number
};

type Props = {
  autoFocus?: boolean,
  excludedUsers: Array<string>,
  handleValueChange: Option => void,
  searchUsers: (string, number) => Promise<*>,
  selectedUser?: Option
};

type State = {
  isLoading: boolean,
  search: string,
  searchResult: Array<Option>
};

const LIST_SIZE = 10;

export default class UsersSelectSearch extends React.PureComponent {
  mounted: boolean;
  props: Props;
  state: State;

  constructor(props: Props) {
    super(props);
    this.handleSearch = debounce(this.handleSearch, 250);
    this.state = { searchResult: [], isLoading: false, search: '' };
  }
  componentDidMount() {
    this.mounted = true;
    this.handleSearch(this.state.search);
  }

  componentWillReceiveProps(nextProps: Props) {
    if (this.props.excludedUsers !== nextProps.excludedUsers) {
      this.handleSearch(this.state.search);
    }
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  filterSearchResult = ({ users }: { users: Array<Option> }) =>
    users.filter(user => !this.props.excludedUsers.includes(user.login)).slice(0, LIST_SIZE);

  handleSearch = (search: string) => {
    this.props
      .searchUsers(search, Math.min(this.props.excludedUsers.length + LIST_SIZE, 500))
      .then(this.filterSearchResult)
      .then(searchResult => {
        if (this.mounted) {
          this.setState({ isLoading: false, searchResult });
        }
      });
  };

  handleInputChange = (search: string) => {
    if (search == null || search.length === 1) {
      this.setState({ search });
    } else {
      this.setState({ isLoading: true, search });
      this.handleSearch(search);
    }
  };

  render() {
    const noResult = this.state.search.length === 1
      ? translateWithParameters('select2.tooShort', 2)
      : translate('no_results');
    return (
      <Select
        autofocus={this.props.autoFocus}
        className="Select-big"
        options={this.state.searchResult}
        isLoading={this.state.isLoading}
        optionComponent={UsersSelectSearchOption}
        valueComponent={UsersSelectSearchValue}
        onChange={this.props.handleValueChange}
        onInputChange={this.handleInputChange}
        value={this.props.selectedUser}
        placeholder=""
        noResultsText={noResult}
        labelKey="name"
        valueKey="login"
        clearable={false}
        searchable={true}
      />
    );
  }
}
