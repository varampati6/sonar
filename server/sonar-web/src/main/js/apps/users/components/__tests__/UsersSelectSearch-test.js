/*
 * SonarQube
 * Copyright (C) 2009-2017 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import React from 'react';
import { shallow } from 'enzyme';
import UsersSelectSearch from '../UsersSelectSearch';

const selectedUser = {
  login: 'admin',
  name: 'Administrator',
  avatar: '7daf6c79d4802916d83f6266e24850af'
};
const users = [
  { login: 'admin', name: 'Administrator', email: 'admin@admin.ch' },
  { login: 'test', name: 'Tester', email: 'tester@testing.ch' },
  { login: 'foo', name: 'Foo Bar', email: 'foo@bar.ch' }
];
const excludedUsers = ['admin'];
const onSearch = jest.fn(() => {
  return Promise.resolve(users);
});
const onChange = jest.fn();

it('should render correctly', () => {
  const wrapper = shallow(
    <UsersSelectSearch
      selectedUser={selectedUser}
      excludedUsers={excludedUsers}
      isLoading={false}
      handleValueChange={onChange}
      searchUsers={onSearch}
    />
  );
  expect(wrapper).toMatchSnapshot();
  const searchResult = wrapper.instance().filterSearchResult({ users });
  expect(searchResult).toMatchSnapshot();
  expect(wrapper.setState({ searchResult })).toMatchSnapshot();
});
