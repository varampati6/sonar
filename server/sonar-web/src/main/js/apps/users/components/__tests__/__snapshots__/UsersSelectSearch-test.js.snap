// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<Select
  addLabelText="Add \\"{label}\\"?"
  arrowRenderer={[Function]}
  autosize={true}
  backspaceRemoves={true}
  backspaceToRemoveMessage="Press backspace to remove {label}"
  className="Select-big"
  clearAllText="Clear all"
  clearValueText="Clear value"
  clearable={false}
  delimiter=","
  disabled={false}
  escapeClearsValue={true}
  filterOptions={[Function]}
  ignoreAccents={true}
  ignoreCase={true}
  inputProps={Object {}}
  isLoading={false}
  joinValues={false}
  labelKey="name"
  matchPos="any"
  matchProp="any"
  menuBuffer={0}
  menuRenderer={[Function]}
  multi={false}
  noResultsText="no_results"
  onBlurResetsInput={true}
  onChange={[Function]}
  onCloseResetsInput={true}
  onInputChange={[Function]}
  openAfterFocus={false}
  optionComponent={[Function]}
  options={Array []}
  pageSize={5}
  placeholder=""
  required={false}
  scrollMenuIntoView={true}
  searchable={true}
  simpleValue={false}
  tabSelectsValue={true}
  value={
    Object {
      "avatar": "7daf6c79d4802916d83f6266e24850af",
      "login": "admin",
      "name": "Administrator",
    }
  }
  valueComponent={[Function]}
  valueKey="login"
/>
`;

exports[`should render correctly 2`] = `
Array [
  Object {
    "email": "tester@testing.ch",
    "login": "test",
    "name": "Tester",
  },
  Object {
    "email": "foo@bar.ch",
    "login": "foo",
    "name": "Foo Bar",
  },
]
`;

exports[`should render correctly 3`] = `
<Select
  addLabelText="Add \\"{label}\\"?"
  arrowRenderer={[Function]}
  autosize={true}
  backspaceRemoves={true}
  backspaceToRemoveMessage="Press backspace to remove {label}"
  className="Select-big"
  clearAllText="Clear all"
  clearValueText="Clear value"
  clearable={false}
  delimiter=","
  disabled={false}
  escapeClearsValue={true}
  filterOptions={[Function]}
  ignoreAccents={true}
  ignoreCase={true}
  inputProps={Object {}}
  isLoading={false}
  joinValues={false}
  labelKey="name"
  matchPos="any"
  matchProp="any"
  menuBuffer={0}
  menuRenderer={[Function]}
  multi={false}
  noResultsText="no_results"
  onBlurResetsInput={true}
  onChange={[Function]}
  onCloseResetsInput={true}
  onInputChange={[Function]}
  openAfterFocus={false}
  optionComponent={[Function]}
  options={
    Array [
      Object {
        "email": "tester@testing.ch",
        "login": "test",
        "name": "Tester",
      },
      Object {
        "email": "foo@bar.ch",
        "login": "foo",
        "name": "Foo Bar",
      },
    ]
  }
  pageSize={5}
  placeholder=""
  required={false}
  scrollMenuIntoView={true}
  searchable={true}
  simpleValue={false}
  tabSelectsValue={true}
  value={
    Object {
      "avatar": "7daf6c79d4802916d83f6266e24850af",
      "login": "admin",
      "name": "Administrator",
    }
  }
  valueComponent={[Function]}
  valueKey="login"
/>
`;
